### PR TITLE
[mlir-tblgen] Relax builder ambiguity check

### PR DESCRIPTION
**Pull request by RoboTux**
The mlir-tblgen tool prevents the parameter of the build() constructor
for the first default-valued attribute of an operation from having a
default value to avoid ambiguity with the corresponding build()
constructor taking unwrapped value. However it does so even when earlier
wrapped unwrappable attribute would lift the ambiguity. This commit
relax the logic accordingly, which allows to remove a manual constructor
in Arith dialect.

### CODE REVIEW

File: mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp [hunk 1]
```diff
@@ -3088,6 +3088,26 @@ void OpEmitter::buildParamList(SmallVectorImpl<MethodParameter> &paramList,
       defaultValuedAttrStartIndex = i;
     }
   }
+
+  // Check if parameters besides default valued one are enough to distinguish
+  // between builders with wrapped and unwrapped arguments.
+  bool hasBuilderAmbiguity = true;
+  for (int i = 0; i < op.getNumArgs(); ++i) {
```
Comment (kuhar) [inline mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:3095]: nit: don't recompute the end after every iteration: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

You could probably also turn this into a range for loop

**Comment by krzysz00** [comment]
Seems reasonable conceptually aside the nits above

**Review by jpienaar — APPROVED** [review]
This looks good, thanks. I'll do a quick check to see if I see any other failures on larger test set later today.

**Comment by jpienaar** [comment]
I see some failures MHLO side ( out-of-line definition of 'build' does not match any declaration ), so seems like this is not just avoiding ambiguity but may also skipping generating some build methods that was generated before.

**Comment by jpienaar** [comment]
Yes, so this results in generating different number of build methods than before in the header, but not the bodies. You can check in build of

tensorflow/compiler/xla/mlir_hlo/_objs/mlir_hlo/hlo_ops

**Review by jpienaar — CHANGES_REQUESTED** [review]
Ensure generating same number of build methods in header as body.

**Review by jpienaar — APPROVED** [review]
OK, I had time to look more. The confusing thing was the user provided builder was not preferred but that's not new here. Post that only 1 change I noticed, seems a different build method being called which resulted in `DefaultValuedOptionalAttr` not being set. But that doesn't seem to be a difference introduce here.

**Pull request merged**